### PR TITLE
New tip : alternatives to classic for loops

### PR DESCRIPTION
**Pull request by sderosiaux**
Simple tip about using Array `forEach`, `filter`, `map` instead of classic for loops.

### CODE REVIEW

**Comment by bbultman** [comment]
While we're on the topic it may be a good idea to also show an example of how they can be composed, too

**Comment by kurtextrem** [comment]
Please include a link to jsperf. These functions aren't faster and currently the tips makes it look like those functions are the better "code style".

**Comment by sderosiaux** [comment]
I don't think talking about performance here is relevant.

Using those functions are not about that at all, but just about the sake of
clarity, simplicity, and good practice. (and it opens perspective to more
concepts such as composition for instance as said previously, and more
generally FP).
I think if you need performance, you will know it and know what to do.

**Comment by kurtextrem** [comment]
Honestly, that is just wrong. For loops have been part of Javascript since ever and EVERY single javascript out there knows how to read it.
Ask a newbie about for loop, what do they do?
Ask a newbie about the "map" function. Where is it from? What does it do? What parameters does it accept? Is there a need to return a value?

**Comment by rocketinventor** [comment]
I agree with @kurtextrem 
These are good functions to know but, saying not to use for loops when possible makes them sound bad when the reality is that they are just not always necessary.

**Comment by manmadareddy** [comment]
+1

**Comment by sderosiaux** [comment]
All right.
I rephrased the content to talk about "alternatives" to classic `for` loops, and not say "avoid them".

**Comment by sderosiaux** [comment]
PR updated.
Now using anonymous functions and added a note about the full functions signature. Thanks.

**Comment by FagnerMartinsBrack** [comment]
`for` loops are imperative, `forEach` is functional. One could argue that imperative programming is better, if one is doing stand-up comedy of course. Clarity and maintainability should be always over early and micro optimizations.

**Comment by homer0** [comment]
- I don't think this is a good tip; these functions have been around for a while, and they are well known to a lot of developers. I think a tip should be something _that you probably didn't know_ (like the first tip or the one about wrapping `use strict`). We should differentiate between tip and tutorial, tutorial being something that explains the standard way to do something.
- You didn't mention something, I think, it's very important to the implementation of `forEach`, `filter` and `map`, and it's the fact that the optional second argument is the value of `this` inside the function (instead of having to use `.bind()` for an object or a class).
- Check tip #0, I think performance it's important here.

Expect the last one, it's all my personal opinion :D.

**Comment by 1robertrb** [comment]
maybe it would be useful to add this example as a reason why you might use forEach instead of for?

``` javascript
[1,2,3].forEach(function(e) {
  doSomethingAsync(function() {
    // access "e"
  });
});
// versus
for(var i = 0; i < arr.length; i++) {
   doSomethingAsync(function() {
      // access arr[i] but it's not what you expect (probably last element).
    });
}
```

that's something i ran into a while ago, i found it useful to know. sorry if its covered somewhere else, i found this repo only lately.

**Comment by sderosiaux** [comment]
@homer0 in my entourage, several didn't used them, for no reason, just because "for i" was doing the job. And you're right about the second arg, indeed ! Thanks. And yes, maybe this is not a "tip" as what is expected on this repo. I'll let @loverajoel decide. (you can close my PR without merging it, I don't mind)
@strcmp Yes indeed, I think everybody ran into this issue. Maybe this is a tip on itself?

**Comment by 1robertrb** [comment]
yeah i think so if it isn't already. maybe it could be written to be more general, i'm most familiar with the `for()` variant but i think it could at least apply to `while` as well.

**Comment by 1robertrb** [comment]
> I don't think this is a good tip; these functions have been around for a while, and they are well known to a lot of developers. 

i don't think this is factual, could you provide evidence of that? that seems to be your impression, not a reality.

**Comment by homer0** [comment]
@strcmp personal experience working with another devs, the fact that they have been around since ECMA 5.1, and, I think the most important, if you go to the tips list, a couple of them make use `.map` and `.forEach` without explaining what they do.

I'm with @chtefi , @loverajoel you have the final word!

**Comment by 1robertrb** [comment]
right but personal experience isn't universal, @chtefi described an experience at odds with yours, and my experience is at odds with yours too.

**Comment by 1robertrb** [comment]
i just feel this way because i feel like javascript is often taught "the wrong way" or people who haven't written any javascript in 10 years, might be unaware that there's an alternative to just using for(), and i think the sighted closure example is a common stumbling block.

**Comment by FagnerMartinsBrack** [comment]
I find interesting a tip that assumes any level of knowledge, but it should never assume a higher level, because those who are not aware will not understand it, while the ones who are aware of it, will just remember it again and maybe help to improve. This is a win win state, any small piece of information increases awareness, it doesn't matter who is reading it.

I believe it should be documented somewhere that **explain something that some people already know is more useful than failing to explain to everyone**.

File: README.md [hunk 1]
```diff
@@ -18,6 +18,84 @@ To get updates, watch the repo and follow the [Twitter account](https://twitter
...
+```javascript
+for (var i = 0; i < elements.length; i++) {
+  var element = elements[i];
+  doSomethingWith(element);
+}
+```
+
+But often, you don't need `i` anywhere else except to get the element at this index.
+You just want to loop over your elements to do something with them.
+An alternative is to use the ES5 Array functions, such as `forEach` :
```
Comment (FagnerMartinsBrack) [inline README.md:36]: Better document an important problem caused by `forEach` which is that you can't break the loop. The alternative is to change to `every` instead when necessary:

``` javascript
var BREAK = false;
var CONTINUE = true;
array.every(function( item ) {
  if ( item.shouldBreak ) {
    return BREAK;
  }
  return CONTINUE;
});
```

Comment (drakmail) [inline README.md:36]: Wow, doesn't know about `every`, it's very useful!

Comment (drakmail) [inline README.md:36]: :+1: for documenting it

**Comment by rocketinventor** [comment]
@strcmp @FagnerMartinsBrack It's unclear from this tip and its thread if you can use the index value "`i`" at all with these methods (`.forEach`, `.map`, etc).

For example, how would you write the following code block using these methods?

```
    //This adds the usenames to the page
    for (var i = 0; i < result.length; i++) {
      document.getElementById("U" + (i + 1)).innerHTML = result[i].user_name;
    }
```

Also, is it possible to multiply each row in one array with the equivalently keyed row in another array in another array or such with these methods (simultaneous forEach with two arrays)?

Thanks!

**Comment by homer0** [comment]
> people who haven't written any javascript in 10 years

Come on xD!! @loverajoel you need to clear this up, what kind of developers are the tips intended for?  noobies, stackoverflowers, advanced, out for 10 years?

**Comment by 1robertrb** [comment]
@homer0 i find it strange how you choose to reply a fragment of everything i said, and ignore everything else. it says a lot. it seems like you are trying hard to make sure people **don't** learn, how bizarre. i am here to teach and help, i have no idea what your purpose is so far. i'll answer your question though, **JSTips are for everyone**, the fact you label people as "noobies", "stackoverflowers", and "advanced" says a lot, i probably won't reply to you again.

@rocketinventor the index is passed as a second argument to forEach, map, & other functions that operate on Arrays, see https://msdn.microsoft.com/en-us/library/ff679980(v=vs.94).aspx as an example.

**Comment by homer0** [comment]
@strcmp **I'm not trying to offend you nor everyone else**; I quoted that because it was really something weird to say; and I was honestly asking @loverajoel ; and complaining about the labels?  I was referencing in part to [this joke](https://www.reddit.com/r/ProgrammerHumor/comments/3dxkka/computer_programming_to_be_officially_renamed/) and, you were never a noobie?, or you never needed stackoverflow?; I have no problem admitting it, I learned about `var` scope a lot after learning js; and a BIG part of my Angular learning came from stackoverflow.

Please don't think I'm against your point of view nor @chtefi tip; please check the first point of my first comment, I don't think the tip is good because it's practically quoting the documentation, just google `forEach javascript` and you'll get the same kind of content. 

Maybe you are right and my question regarding "what kind of developers are the tips indented for?" should've been "what's the type of tips you expect?", a tip that it's practically documentation about the language is a good tip?

Again, sorry if you thought I was trying to start a fight, the labels part was wrong, but you have to agree, the `10 years` part was funny :stuck_out_tongue_closed_eyes:

**Comment by sderosiaux** [comment]
Well, I think this "tip" has gone too far and is too wide, too tutorialish yes (too many details to cover, look at all those comments, for one "tip"). It should be more focused on a particular feature, such as the other tips.
@homer0 "just google forEach javascript" : I think this repo is useful to make you learned/realized something you would not look for explicitely.

**Comment by 1robertrb** [comment]
@homer0 :laughing: okay, no problem i guess it was funny. i don't really look at stuff that way (noobie, etc), you're homer0 and your words are what i pay attention to. thats it really. with the 10 year reference i'm trying to say 10 years ago javascript was a lot different and people might have learnt it then and come away with the wrong impression. 

maybe tips like this would help people rediscover new things about javascript, avoid frustrating stumbling blocks like the example and so on :)

**Comment by 1robertrb** [comment]
@rocketinventor  "For example, how would you write the following code block using these methods?"

``` javascript
    // Original method (for)
    for (var i = 0; i < result.length; i++) {
      document.getElementById("U" + (i + 1)).innerHTML = result[i].user_name;
    }
    // New method (forEach)
    result.forEach(function(element, i) {
      document.getElementById("U" + (i + 1)).innerHTML = element.user_name;
    });
```

@rocketinventor "Also, is it possible to multiply each row in one array with the equivalently keyed row in another array in another array or such with these methods (simultaneous forEach with two arrays)?"

``` javascript
// pirates.js
var arr1 = [1,2,3,4],
      arr2 = [5,6,7,8];
arr1.map(function(_, index) {
  return arr1[index] * arr2[index];
});
```

i didnt test it, meant to give you a general idea.

**Comment by FagnerMartinsBrack** [comment]
@chtefi 

> It should be more focused on a particular feature, such as the other tips.

He has a point.
What about documenting that a tip is valid while it doesn't take `x` minutes to read the content and links, being `x` a value previously defined?

Does it makes sense creating an issue about it?

**Comment by loverajoel** [comment]
What a nice conversation!
I don't know what kind of developers are the tips intended for, it's the same conversation if ES6 simple functionality is a tip or not, for people that never read about it is a tip, same here.
I'm pretty sure that there are lot people that are starting with js and don't know about `map()` or `filter()`.

I haven't the final word, this is of the community and always I waited for each PR comments from the community before merge. (this is the thread most huge :p)

There 360 days and 360 tips per year, sufficient space for all kind of tips. Maybe in the while we can review this and the destination of the project!

**Comment by loverajoel** [comment]
Hi, how you can see the jstips repository changed his format and the `master` branch now is `gh-pages`, now are based in _jekyll blog format_. You can see live here => [jstips.com](http://www.jstips.co/)
For that reason you should update your tip with this [instructions](https://github.com/loverajoel/jstips/blob/gh-pages/CONTRIBUTING.md).

Let me know if you need some help or have suggestion. All are welcome!

Thanks

**Pull request closed without merging**